<a href="https://colab.research.google.com/github/worawit-saetan/Fchollet_Book_Deep_Learning_with_Python/blob/main/Chapter_2_MNIST_dataset_in_Keras.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

This is the code from François Chollet Book - Chapter 2 The MNIST dataset in Keras

I write and run the code to better understand the Chapter 2

Loading the MNIST dataset in Keras

In [1]:
from keras.datasets import mnist
(train_images, train_labels), (test_images, test_labels) = mnist.load_data()

Checking the training and test dataset

In [2]:
train_images.shape


(60000, 28, 28)

In [3]:
train_labels

array([5, 0, 4, ..., 5, 6, 8], dtype=uint8)

In [4]:
len(train_labels)

60000

In [5]:
test_images.shape

(10000, 28, 28)

In [6]:
test_images.shape

(10000, 28, 28)

In [7]:
test_labels

array([7, 2, 1, ..., 4, 5, 6], dtype=uint8)

The network architecture with Sequential layers

2 Dense layers with 512 nodes and 10 nodes(for classifying 0-9 digits)

In [8]:
from keras import models
from keras import layers
network = models.Sequential()
network.add(layers.Dense(512, activation='relu', input_shape=(28*28,)))
network.add(layers.Dense(10, activation='softmax'))

The compilation step 

In [9]:
network.compile(optimizer='rmsprop',loss='categorical_crossentropy',metrics=['accuracy'])

Preparing the image data : reshape and rescale the training/test dataset

In [10]:
train_images = train_images.reshape((60000, 28 * 28))
test_images = test_images.reshape((10000, 28 * 28))
print(train_images.shape)
print(test_images.shape)

(60000, 784)
(10000, 784)


In [11]:
train_images = train_images.astype('float32') / 255
test_images = test_images.astype('float32') / 255

Preparing the labels : encode the labels of the training dataset and test set to categorical

In [12]:
from keras.utils import to_categorical
train_labels = to_categorical(train_labels)
test_labels = to_categorical(test_labels)

In [13]:
print(train_labels.shape)
print(train_labels)

(60000, 10)
[[0. 0. 0. ... 0. 0. 0.]
 [1. 0. 0. ... 0. 0. 0.]
 [0. 0. 0. ... 0. 0. 0.]
 ...
 [0. 0. 0. ... 0. 0. 0.]
 [0. 0. 0. ... 0. 0. 0.]
 [0. 0. 0. ... 0. 1. 0.]]


In [14]:
print(test_labels.shape)
print(test_labels)

(10000, 10)
[[0. 0. 0. ... 1. 0. 0.]
 [0. 0. 1. ... 0. 0. 0.]
 [0. 1. 0. ... 0. 0. 0.]
 ...
 [0. 0. 0. ... 0. 0. 0.]
 [0. 0. 0. ... 0. 0. 0.]
 [0. 0. 0. ... 0. 0. 0.]]


Train the network : 5 times with batch size = 128 

In [15]:
network.fit(train_images, train_labels, epochs=5, batch_size=128)

Epoch 1/5
469/469 [==============================] - 6s 11ms/step - loss: 0.2546 - accuracy: 0.9269
Epoch 2/5
469/469 [==============================] - 5s 11ms/step - loss: 0.1034 - accuracy: 0.9699
Epoch 3/5
469/469 [==============================] - 5s 11ms/step - loss: 0.0678 - accuracy: 0.9798
Epoch 4/5
469/469 [==============================] - 5s 11ms/step - loss: 0.0501 - accuracy: 0.9851
Epoch 5/5
469/469 [==============================] - 5s 11ms/step - loss: 0.0379 - accuracy: 0.9888


The training set accuracy is 0.988 (98.8%)

Let's check on the test set

In [16]:
network.evaluate(test_images,test_labels)

313/313 [==============================] - 1s 3ms/step - loss: 0.0684 - accuracy: 0.9793


[0.06836093962192535, 0.9793000221252441]

The test set accuracy turns out to be 97.9%

that’s quite a bit lower than the training
set accuracy. This gap between training accuracy and test accuracy is an example of overfitting (the fact that machine learning models tend to perform better on training data than new data.

In [17]:
train_labels.dtype

dtype('float32')